### ISSUE TITLE
Index Checker: Print either required: "integer>=0" or "@NonNegative or @Positive" but not both

### ISSUE THREAD

**Issue by Kakarot-SSJ4** [typetools/checker-framework#2344]
Consider the following code:
```
import java.util.*;
class errormsg{
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int i = sc.nextInt();
		int ar[] = new int[i];
	}
}
```
Type in the terminal: javac -processor org.checkerframework.checker.index.IndexChecker errormsg.java
It produces the following error: 
```
errormsg.java:6: error: [array.length.negative] Variable used in array creation could be negative.
		int ar[] = new int[i];
		                   ^
  found   : @LowerBoundUnknown int
  required: an integer >= 0 (@NonNegative or @Positive)
1 error
```
One of the following: "an integer>=0" or "@NonNegative or @Positive" is redundant and should not be displayed for less clustered error details.

**Comment by kelloggm** [comment]
Thanks for the report! The rationale for the error message including both is that a variable with either type is acceptable. You're right, though, that `@Positive` is a subtype of `@NonNegative`, so the error message is mostly redundant.

I think it would be okay to write the error message so that `@Positive` is not mentioned, but I also don't think there's anything wrong with the error message that's currently issued.

Is there evidence that the current error message is confusing for users? I don't think it's worth investing in the change unless we have a compelling reason that shows us that something is wrong.

**Comment by Kakarot-SSJ4** [comment]
You misunderstood my interpretation, I meant to convey that either "an integer >= 0" be printed or "@NonNegative or @Positive"(both together) be printed, as either of the above two options completely describe the requirement.

**Comment by kelloggm** [comment]
Oh! My apologies.

You're right that either would be sufficient, but there is a significant advantage to having both:
- `"an integer >= 0"` is in the language of mathematics; a programmer does not need to know anything about the typechecker to understand what it means
- `@NonNegative or @Positive` is in the language of the typechecker. Part of the job of the error message is to educate the reader about what the tool expects.

In other words, the first part of the error message explains the error in standard English, while the second part connects that English to the annotations that the user could write to eliminate the warning.
